scarlet/initialization.py

```diff
@@ -834,157 +433,181 @@ def initSource(
     fallback : bool
         Whether to reduce the number of components
-        if the model cannot be initialized with `maxComponents`.
+        if the model cannot be initialized with `max_components`.
         This is unlikely to be used in production
         but can be useful for troubleshooting when an error can cause
         a particular source class to fail every time.
     """
-    from .source import PointSource, ExtendedSource
+    from .source import ExtendedSource
 
-    while maxComponents > 1:
-        try:
-            source = ExtendedSource(
-                frame,
-                center,
-                observation,
-                thresh=thresh,
-                shifting=shifting,
-                K=maxComponents,
-            )
-            try:
-                source.check_parameters()
-                # Make sure that SED is >0 in at least 1 band
-                if np.any(
-                    [
-                        np.all(child.children[0].get_model() <= 0)
-                        for child in source.children
-                    ]
-                ):
-                    raise ArithmeticError
-            except ArithmeticError:
-                msg = "Could not initialize source at {} with {} components".format(
-                    center, maxComponents
-                )
-                logger.warning(msg)
-                raise ValueError(msg)
-
-            if downgrade and np.all(np.array(source.bbox.shape[1:]) <= 8):
-                # the source is in a small box so it must be a point source
-                maxComponents = 0
-            elif downgrade and np.all(np.array(source.bbox.shape[1:]) <= 16):
-                # if the source is in a slightly larger box
-                # it is not big enough to model with 2 components
-                maxComponents = 1
-            elif hasEdgeFlux(source, edgeDistance):
-                source.shifting = True
-
-            break
-        except Exception as e:
-            if not fallback:
-                raise e
-            # If the MultiComponentSource failed to initialize
-            # try an ExtendedSource
-            maxComponents -= 1
+    if not hasattr(observations, "__iter__"):
+        observations = (observations,)
 
-    if maxComponents == 1:
-        try:
-            source = ExtendedSource(
-                frame, center, observation, thresh=thresh, shifting=shifting
-            )
+    source_shifting = shifting
+    _, psf_snr = get_psf_spectrum(center, observations, compute_snr=True)
 
-            try:
-                source.check_parameters()
-                if np.all(source.children[0].get_model() <= 0):
-                    raise ArithmeticError
-            except ArithmeticError:
-                msg = "Could not initlialize source at {} with 1 component".format(
-                    center
+    while max_components >= 0:
+        try:
+            if np.abs(psf_snr) < min_snr * max_components:
```

Comment:
I agree with the first suggestion, and will implement it.

The second suggestion is a different approach. So far, we have iteratively reduced the number of components. This would suggest a number of components instead. I think our approach is OK as long as K=2.

Reply:
I don't see how the approach is different. The current algorithm will start with max_components, check to see if the psf_snr is large enough for that number of components, fail if it isn't, and iterate that process until max_components is small enough that psf_snr < min_snr * max_components.

My suggestion does the exact same thing, except it doesn't have to go through the try/except block to decrease max_components, it just sets it to the minimum viable number of components that make your condition true. Note that this will never be larger than max_components, so it does exactly the same thing.

Reply:
changed